sql/cache.go

```diff
@@ -131,7 +131,11 @@ func (m mapCache) Put(u uint64, i interface{}) error {
 }
 
 func (m mapCache) Get(u uint64) (interface{}, error) {
-	return m.cache[u], nil
+	v, ok := m.cache[u]
+	if !ok {
+		return nil, ErrKeyNotFound.New(u)
```

Comment:
This has a lot of performance overhead. Do we need this to distinguish null entries in the cache?

Reply:
This was needed to match the existing behavior of the `lruCache` and `historyCache` implementations. They both return `ErrKeyNotFound` when the key isn't found, and we have some other code (in Subquery.Eval if I remember correctly) that depends on that behavior to work correctly. I believe we can trigger a bug with the current code today under the right situation, since it won't detect nil as not found.

Reply:
If perf from looking at the error response is a concern, it of course probably wouldn't be too hard to change the API expectations to not return that error.

Reply:
I dug through usage of `KeyValueCache.Get` some more this morning and there are several places where the code does need to distinguish between cached with a result of nil and not present. I think we still want this change to make the `mapCache` implementation work the same as the other implementations, but let me know if you think refactoring the interface for the cache implementations to not use an error for this signal needs to happen as part of this PR.